File: January 2026 GFG SOLUTION/January-22.py
class Solution:
    def subarrayRanges(self, a):
        n = len(a)
        res = 0
        s = []
        for i in range(n + 1):
            while s and (i == n or a[s[-1]] > a[i]):
                j = s.pop()
                k = s[-1] if s else -1
                res -= a[j] * (i - j) * (j - k)
            s.append(i)
        s.clear()
        for i in range(n + 1):
            while s and (i == n or a[s[-1]] < a[i]):
                j = s.pop()
                k = s[-1] if s else -1
                res += a[j] * (i - j) * (j - k)
            s.append(i)
        return res

File: January 2026 GFG SOLUTION/January-22.java
class Solution {
    public int subarrayRanges(int[] a) {
        int n = a.length;
        long res = 0;
        Stack<Integer> s = new Stack<>();
        for (int i = 0; i <= n; i++) {
            while (!s.isEmpty() && (i == n || a[s.peek()] > a[i])) {
                int j = s.pop();
                int k = s.isEmpty() ? -1 : s.peek();
                res -= (long)a[j] * (i - j) * (j - k);
            }
            s.push(i);
        }
        s.clear();
        for (int i = 0; i <= n; i++) {
            while (!s.isEmpty() && (i == n || a[s.peek()] < a[i])) {
                int j = s.pop();
                int k = s.isEmpty() ? -1 : s.peek();
                res += (long)a[j] * (i - j) * (j - k);
            }
            s.push(i);
        }
        return (int)res;
    }
}
